Stop using dynamic keys with the database cache

RadoslavGeorgiev opened the pull request:
Fixes https://github.com/Automattic/woocommerce-payments/issues/10617

#### Changes proposed in this Pull Request

- [x] Remove the dynamic keys from `WC_Payments_Onboarding_Service` in favor of storing the locale (& country code) within the array. This is done under the assumption that those change, but not frequently.
- [x] As in the original issue, change where customer payment methods are stored to storage that has cleanup / garbage collection mechanisms. I choose to only cache payment methods for logged-in users, and to use a single and easily traceable user meta entry for cache.

The latter is the bigger part of this PR, and replaces the current storage of payment methods as options with dynamic keys to user meta under a single key (`_wcpay_payment_methods`). The stored value looks like this:

```json
{
	"customer_id": "cus_TEwTgNq3kg25ff",
	"payment_methods_woocommerce_payments": [
		{ "payment_method" },
		{ "payment_method" }
	]
}
``` 

Upon follow-up calls, we validate that the customer ID is the same as expected and if so, use the cached value.

Considerations:

1. There are some APIs (REST & extensibility) having methods that can fetch payment method details, and those will not be cached. However, I am not aware of anything that uses those APIs, we built them for extensibility, but have no feedback of them being used.
2. There will be no caching for non-logged-in customers. However, there are no calls to fetch stored payment methods for them either.
3. The gateway ID is included in the cache to support legacy functionality like SEPA, which is a separate gateway. I doubt that's available anymore, but I wanted to preserve backward compatibility.

#### Testing instructions

1. Onboard a new merchant, make sure that everything looks well (available payment methods, etc.)
2. Check out and store your payment method, either as a logged-in or logged-out user.
3. Repeat a couple of times to have a few payment methods.
4. Navigate to My Account and delete either payment method. It should disappear as expected.

I recommend connecting to xDebug within `get_payment_methods_from_stripe()` to confirm whether the cache is hit, busted, or updated properly.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

RadoslavGeorgiev committed:
Remove unused force refresh flags

RadoslavGeorgiev committed:
For recommended payment methods & onboarding data, store country code & locale in the cached value, rather than different prefix.

RadoslavGeorgiev committed:
Changelog entry

RadoslavGeorgiev committed:
Remove the onboarding and payment method keys as prefixes

RadoslavGeorgiev committed:
Remove a left-over key for payment processing factors

RadoslavGeorgiev committed:
Transfer functionality to the token service, delete the rest

Some functionality (ex. clearing payment methods) was moved to the token service, and the service was added where needed.

This left the database cache clear of all methods that have dynamic keys.

The token service still needs cache and some improvements.

RadoslavGeorgiev committed:
Remove the webhook processing service's dependency

RadoslavGeorgiev committed:
Add user-based caching to the token service, fix the methods that clear the cache

RadoslavGeorgiev committed:
Fix most existing tests

RadoslavGeorgiev committed:
Include gateway IDs in payment method cache. Remove calls to delete from the DB cache with prefix

RadoslavGeorgiev committed:
Add tests for the new methods of the token service

RadoslavGeorgiev committed:
Clean up properly after token service tests

RadoslavGeorgiev committed:
Merge branch 'develop' into WOOPMNT-4879

RadoslavGeorgiev committed:
Remove todo comment, caching is now in place

leonardola commented on includes/class-wc-payments-token-service.php:
```diff
@@ -210,6 +243,63 @@ public function woocommerce_get_customer_payment_tokens( $tokens, $user_id, $gat
+		if (
+			is_array( $cached_data )
+			&& (
+				( isset( $cached_data['customer_id'] ) && $cached_data['customer_id'] !== $customer_id )
```
This means that the same user could have multiple $customer_ids. When would this scenario happen? And when it happens shouldn't we keep both cached?

RadoslavGeorgiev replied:
This would happen when there was one ID, associated with a given merchant account, and then the account is changed. So if there was a customer ID in test mode, there would be none in live mode and vice-versa. Or if the merchant got reonboarded, for example. That's why the account update webhook flushes all the cache.

That's why it makes sense to only keep one customer ID, and drop everything that existed already in case that customer ID changes.

leonardola commented on tests/unit/test-class-wc-payments-token-service.php:
```diff
@@ -799,4 +807,298 @@ private function generate_link_token( $stripe_id, $wp_id = 0 ) {
+		update_user_meta( $user_id, $cache_key, $cached_data );
+
+		// Verify cached data exists.
+		$this->assertEquals( $cached_data, get_user_meta( $user_id, $cache_key, true ) );
```
Not truly necessary. It's just testing if WordPress works which we should take for granted in this case IMHO

RadoslavGeorgiev replied:
Good point, removed in https://github.com/Automattic/woocommerce-payments/pull/11092/commits/6eb011e0e031a520ce66b2bdd31749ee19999367

leonardola commented on tests/unit/test-class-wc-payments-token-service.php:
```diff
@@ -799,4 +807,298 @@ private function generate_link_token( $stripe_id, $wp_id = 0 ) {
+		update_option( 'wcpay_other_option', 'should_not_be_deleted' );
+
+		// Verify cached data exists.
+		$this->assertEquals( $cached_data_1, get_user_meta( $user_id_1, WC_Payments_Token_Service::CACHED_PAYMENT_METHODS_META_KEY, true ) );
```
once again probably not necessary

RadoslavGeorgiev replied:
Also removed in https://github.com/Automattic/woocommerce-payments/pull/11092/commits/6eb011e0e031a520ce66b2bdd31749ee19999367

leonardola commented on tests/unit/test-class-wc-payments-token-service.php:
```diff
@@ -799,4 +807,298 @@ private function generate_link_token( $stripe_id, $wp_id = 0 ) {
+		add_filter( 'wcpay_force_network_saved_cards', '__return_true' );
+
+		// Clear all cached payment methods.
+		$this->token_service->clear_all_cached_payment_methods();
```
other than this function call the rest of this test is pretty similar to `test_clear_cached_payment_methods_for_user_network_saved_cards_enabled` is there a way to refactor this test case into a more re-usable form?

RadoslavGeorgiev replied:
Yes. In https://github.com/Automattic/woocommerce-payments/pull/11092/commits/4dcd3e03444781aa21138985730d9a07117b98b8 I combined them into a single test with a data provider.

leonardola commented on tests/unit/test-class-wc-payments-token-service.php:
```diff
@@ -799,4 +807,298 @@ private function generate_link_token( $stripe_id, $wp_id = 0 ) {
+		$this->assertEquals( $payment_methods, $cached_data[ 'payment_methods_' . $gateway_id ] );
+	}
+
+	private function call_sut_method( $method_name, $user_id, $customer_id, $gateway_id ) {
```
what does sut mean in this context?

RadoslavGeorgiev replied:
Service under test. It's frequently used for newer tests in core, and I just got used to it.

When I write new test classes, I typically use `$sut` to store the main class that is being tested. Then I don't have to wonder how the property is called for different classes, `$this->sut` is always there. :)

leonardola commented on tests/unit/test-class-wc-payments-token-service.php:
```diff
@@ -799,4 +807,298 @@ private function generate_link_token( $stripe_id, $wp_id = 0 ) {
+		$this->assertEquals( $cached_payment_methods, $result );
+
+		// Verify customer service is not called (since we're using cached data).
+		$this->mock_customer_service
```
Doesn't this need to be moved to a line before we try to call the method?

RadoslavGeorgiev replied:
Yes 🙈 

Fixed in https://github.com/Automattic/woocommerce-payments/pull/11092/commits/c58952aba11edc179a73a71a7e295e133056d12b

leonardola commented on tests/unit/test-class-wc-payments-token-service.php:
```diff
@@ -799,4 +807,298 @@ private function generate_link_token( $stripe_id, $wp_id = 0 ) {
+	private function call_sut_method( $method_name, $user_id, $customer_id, $gateway_id ) {
+		// Use reflection to access private method.
+		$reflection = new ReflectionClass( $this->token_service );
+		$method     = $reflection->getMethod( 'get_payment_methods_from_stripe' );
```
we're always calling `get_payment_methods_from_stripe` should that be calling `$method_name` instead or should we remove the `$method_name` parameter?

RadoslavGeorgiev replied:
This ended up being a left-over. The point of the private method is to be reusable, so I started using `$method_name` in https://github.com/Automattic/woocommerce-payments/pull/11092/commits/6eb011e0e031a520ce66b2bdd31749ee19999367.

leonardola commented on tests/unit/test-class-wc-payments-token-service.php:
```diff
@@ -799,4 +807,298 @@ private function generate_link_token( $stripe_id, $wp_id = 0 ) {
+				[ $payment_methods[1] ]
+			);
+
+		$result = $this->call_sut_method( 'get_payment_methods_for_customer', $user_id, $customer_id, $gateway_id );
```
Why are we calling a method that was just mocked?

RadoslavGeorgiev replied:
I changed the method name in https://github.com/Automattic/woocommerce-payments/pull/11092/commits/6eb011e0e031a520ce66b2bdd31749ee19999367, it was always supposed to be `get_payment_methods_from_stripe`. However, even before that, I was calling the `get_payment_methods_for_customer` of the token service, while mocking the method of the customer service. Same name, different methods.

leonardola commented on tests/unit/test-class-wc-payments-token-service.php:
```diff
@@ -799,4 +807,298 @@ private function generate_link_token( $stripe_id, $wp_id = 0 ) {
+			->with( $customer_id, Payment_Method::SEPA )
+			->willReturn( $payment_methods );
+
+		$result = $this->call_sut_method( 'get_payment_methods_for_customer', $user_id, $customer_id, $gateway_id );
```
once again calling a method that was just mocked

RadoslavGeorgiev replied:
Same as https://github.com/Automattic/woocommerce-payments/pull/11092#discussion_r2434842884. It was both the wrong method (though different from the mocked one) and the parameter was not used. Fixed in https://github.com/Automattic/woocommerce-payments/pull/11092/commits/6eb011e0e031a520ce66b2bdd31749ee19999367

leonardola commented on changelog/WOOPMNT-4879:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: dev
+
+Stop using dynamic keys for the database cache in favor of better validation and transients.
```
The entry log mentions transients but `Database_Cache` still uses `update_option`

RadoslavGeorgiev replied:
For once I decided to add the changelog entry in the initial commits of the PR, and this happened! 🤣

I'll update it.

RadoslavGeorgiev replied:
Updated in https://github.com/Automattic/woocommerce-payments/pull/11092/commits/8c6cc6fa5e51374248a1a010d21d2c69552e686e to 

```
Stop using dynamic keys for the database cache and move cached payment methods to user meta.
```

leonardola reviewed (commented):
My understanding is that we only remove cached payment methods when there is an merchant account update, [the customer deletes a payment method](https://github.com/Automattic/woocommerce-payments/blob/53fea0a137af7f518e74b11471d4acb02c7cf897/includes/class-wc-payments-token-service.php#L382), [the customer adds a new payment method](https://github.com/Automattic/woocommerce-payments/blob/53fea0a137af7f518e74b11471d4acb02c7cf897/includes/class-wc-payments-token-service.php#L74) or [the customer sets the default payment method](https://github.com/Automattic/woocommerce-payments/blob/53fea0a137af7f518e74b11471d4acb02c7cf897/includes/class-wc-payments-token-service.php#L415). 

Do you think it would be beneficial to add a scheduled action to proactively remove stale cached methods? reducing the size of the `wp_usermeta` table?

RadoslavGeorgiev commented:
Thanks for the review @leonardola!

> My understanding is that we only remove cached payment methods when there is an merchant account update, [the customer deletes a payment method](https://github.com/Automattic/woocommerce-payments/blob/53fea0a137af7f518e74b11471d4acb02c7cf897/includes/class-wc-payments-token-service.php#L382), [the customer adds a new payment method](https://github.com/Automattic/woocommerce-payments/blob/53fea0a137af7f518e74b11471d4acb02c7cf897/includes/class-wc-payments-token-service.php#L74) or [the customer sets the default payment method](https://github.com/Automattic/woocommerce-payments/blob/53fea0a137af7f518e74b11471d4acb02c7cf897/includes/class-wc-payments-token-service.php#L415). 
> 
> Do you think it would be beneficial to add a scheduled action to proactively remove stale cached methods? reducing the size of the `wp_usermeta` table?

Yes and no. It could potentially be beneficial, but I'd say that the effect would be rather negligible. The way metadata is loaded (altogether at the first load of the object, user in this case), up to a single entry per user does not constitute a significant slow-down in any case.

On the other hand, even with just Card and Link enabled, not having the data handy, slows down the initial checkout page by more than a second for me. I think this is worth the sacrifice.

---

Regarding tests, I might have decided to over-utilize Cursor at the end of my work day. I'll optimize them first thing tomorrow.

RadoslavGeorgiev committed:
Remove redundant conditions from tests, fix the call for internal methods

RadoslavGeorgiev committed:
Update the changelog entry

RadoslavGeorgiev committed:
Move a never() mock before the method call

RadoslavGeorgiev committed:
Combine tests for network cards

leonardola commented on includes/class-wc-payments-onboarding-service.php:
```diff
@@ -908,12 +919,9 @@ public function cleanup_on_account_reset() {
 	 */
 	public function cleanup_on_account_onboarded() {
```
I don't see this method being called when I try to re-onboard on my local environment.
Steps taken:
- Store is already onboarded
- Go to /wp-admin > WCPay dev > Go to WCPay Onboarding > Reset Account
- Then I setup a new test account
- The setup request finishes successfully. But the loader never completes so I get stuck there
- This method is not called
- My new account gets stuck in Pending Approval so I have to manually clean the WooPayments cache via WCPay Dev

RadoslavGeorgiev replied:
I think this might be my fault! 😨 

Not something from within this PR, but this seems to be an edge case for account mirroring to Transact Platform V2. I'll fix that separately, but in the interim, I suggest commenting [this line](https://github.com/Automattic/transact-platform-server/blob/73f7a2a681af93e7d1ddb64e4e0980bfbe68f1a8/transact-platform-server.php#L69) out. Once you do, the process should proceed as usual.

Keep in mind that the method will not be called until you go through the embedded onboarding, meaning that the test-drive account (where you don't go through any Stripe pages) is not sufficient to reach the method.

leonardola commented on includes/class-wc-payments-token-service.php:
```diff
@@ -210,6 +243,63 @@ public function woocommerce_get_customer_payment_tokens( $tokens, $user_id, $gat
+		if (
+			is_array( $cached_data )
+			&& isset( $cached_data[ $cache_key ] )
+			&& is_array( $cached_data[ $cache_key ] )
```
When I visit /my-account > Payment Methods the `$cache_key` is set to `payment_methods_` instead of `payment_methods_woocommerce_payments` so the cache is not used for the first page load.

Steps followed:
- Remove the `_wcpay_payment_methods` meta from the database
- Add breakpoint [here](https://github.com/Automattic/woocommerce-payments/blob/4dcd3e03444781aa21138985730d9a07117b98b8/includes/class-wc-payments-token-service.php#L254)
- Visit /my-account > Payment Methods
- First time the function is called the cache does not exist
- Second time it does but uses the wrong key so it has to recreate the cache again
- Same happens when I change the default payment method

RadoslavGeorgiev replied:
Great catch! We are using an undocumented [filter from core](https://github.com/woocommerce/woocommerce/blob/69dc20de82080a86ff1c523c455f22ca908f432f/plugins/woocommerce/includes/class-wc-payment-tokens.php#L76) and [assuming that the gateway ID is always a string](https://github.com/Automattic/woocommerce-payments/blob/97d154041eb7f39cc6724e2f74a28879699fcb86/includes/class-wc-payments-token-service.php#L183-L186), even though that is not the case.

Seems like this will require me to rewrite the `get_payment_methods_from_stripe` method and particularly the format of the cached data. Right now, it follows this format:

```js
{
	"customer_id": "cus_XYZ",
	"payment_methods_{gateway_id}": [
		// payment methods here
	]
}
```

However, with the gateway ID possibly missing, I can't use it. I'll switch back to using the `$type` from `$retrievable_payment_method_types` instead of the gateway ID. The data would look pretty similar with `{type}` instead of `{gateway_id}` in the key, but instead of checking the cache and returning it at once, I'll always go through the foreach loop for each payment method and check the cache there.

I'll implement this on Monday, but I'm curious what you think in the meantime.

leonardola replied:
I'm not sure which option is the fastest but caching something with the wrong name is prone to other problems. Let's try the loop version.

RadoslavGeorgiev replied:
I updated the caching strategy in https://github.com/Automattic/woocommerce-payments/pull/11092/commits/554e866514704a4d51f0c1c302298cdd2912c3ea

Now instead of using the gateway ID as a part of the cache key, just like before the PR, it is now used to load the necessary payment method types and only those are keyed in the array.

leonardola commented on includes/class-wc-payments-token-service.php:
```diff
@@ -133,6 +135,46 @@ public function is_valid_payment_method_type_for_gateway( $payment_method_type,
+	 * Clear all cached payment methods.
+	 * Used when account data is updated and all payment method caches need to be cleared.
+	 */
+	public function clear_all_cached_payment_methods() {
```
I updated the merchant account via stripe dashboard. The `account.updated` event got to the transact server but this function was never called. What's the proper way to trigger this flow?

Steps taken:
- Visit Stripe account dashboard
- Make sure your transact server is listening to Stripe webhooks
- Add a breakpoint [here](https://github.com/Automattic/woocommerce-payments/blob/4dcd3e03444781aa21138985730d9a07117b98b8/includes/class-wc-payments-token-service.php#L156)
- Update the `Personal details`
- Notice the function is never called

RadoslavGeorgiev replied:
The `account.updated` webhook for the client seems not to be a mirror image of Stripe's `account.updated`, nor sent at the same time.

I assume that you were expecting this to happen through the changed webhook processing service. However, after [this PR](https://github.com/Automattic/transact-platform-server/pull/7653) (very recent), the `account.updated` seems to be sent upon specific conditions, ex. hitting some account-based limit. 

I added https://github.com/Automattic/woocommerce-payments/pull/11092/commits/97d154041eb7f39cc6724e2f74a28879699fcb86 to make sure that we use the new `account.deleted` webhook to clean up payment methods.

Now, you should be able to hit the breakpoint shortly after starting the account reset process, even with a test drive account.

leonardola replied:
I don't see the `account.deleted` webhook being triggered after reseting and re-onboarding the test account and it didn't hit the breakpoint either.

<img width="913" height="902" alt="Screenshot 2025-10-22 at 15 58 12" src="https://github.com/user-attachments/assets/bf6d1959-22be-45ee-b5d0-d56629f8c1b2" />

RadoslavGeorgiev replied:
I just tried this, and a simple change of the street number of the connected account triggered an `account.updated` webhook to the server, and that in turn triggered `clear_all_cached_payment_methods()`.

I _think_ that you might not be receiving webhooks because the server sends them asynchronously now, and that requires CRON. However, WP CRON might be deactivated for you. Try running `npm run cron:stop` (or `:start`, it's weird) within the server repo before trying again.

mgascam replied:
It also worked well for me while testing locally. I changed the "Support email" in Payments Settings and the `account.updated `was triggered and `clear_all_cached_payment_methods` was called. It correctly deleted the cached payment methods from `wp_usermeta` and `wp_options` .

leonardola reviewed (commented):
I tried different methods to trigger the removal of the `wp_options` data but it never happened. See comments below

RadoslavGeorgiev committed:
Clear cached payment methods upon the account.deleted webhook

leonardola commented on includes/class-wc-payments-onboarding-service.php:
```diff
@@ -124,23 +124,16 @@ public function init_hooks() {
 	 * @return ?array Fields data, or NULL if failed to retrieve.
 	 */
-	public function get_fields_data( string $locale = '', bool $force_refresh = false ): ?array {
+	public function get_fields_data( string $locale = '' ): ?array {
```
This is public method. I'm not sure about the backward compatibility policy for WooPayments but any other 3rd party plugin using this method will break after this change

RadoslavGeorgiev replied:
I added back the parameter together with a deprecation warning in https://github.com/Automattic/woocommerce-payments/pull/11092/commits/524b10647c07f800d057f097daef363223b17d99.

leonardola commented on includes/class-wc-payments-onboarding-service.php:
```diff
@@ -167,23 +169,38 @@ function () use ( $locale ) {
+					$payment_methods = $this->payments_api_client->get_recommended_payment_methods( $country_code, $locale );
+
+					// Indicate that the cached value is specific for the given locale and country code.
+					return [
```
Changing the returned data also breaks backward compatibility

RadoslavGeorgiev replied:
Actually, that's not the case. The stored format is changed here:

https://github.com/Automattic/woocommerce-payments/blob/524b10647c07f800d057f097daef363223b17d99/includes/class-wc-payments-onboarding-service.php#L184-L188

However, the important piece is `payment_methods` and only that gets returned from the method:

https://github.com/Automattic/woocommerce-payments/blob/524b10647c07f800d057f097daef363223b17d99/includes/class-wc-payments-onboarding-service.php#L208

`__locale` and `__country_code` are not used or leaked outside the method.

leonardola reviewed (commented):
Overall the changes look good. Left a few comments about backward compatibility and the old options still not being deleted

RadoslavGeorgiev committed:
Merge branch 'develop' into WOOPMNT-4879

RadoslavGeorgiev committed:
Switch PM caching strategy

Instead of using the gateway ID for cache, use the gateway ID to check what payment method types should be cached. Only load the unavailable ones.

RadoslavGeorgiev committed:
Fix tests

RadoslavGeorgiev committed:
Maek the `$force_refresh` parameter as deprecated.

RadoslavGeorgiev commented:
Requesting a review from Gamma as well, as the PR has been stale for a while, and Leonardo might have lost context already.

mgascam commented on includes/class-database-cache.php:
```diff
@@ -86,6 +70,26 @@ class Database_Cache implements MultiCurrencyCacheInterface {
+	 *
+	 * @var string[]
+	 */
+	const ALL_KEYS = [
```
The `ADDRESS_AUTOCOMPLETE_JWT_KEY` constant is missing here, just checking if it is intentional?

RadoslavGeorgiev replied:
Not at all. I think it might have been added after I started the PR. With https://github.com/Automattic/woocommerce-payments/pull/11092/commits/38cb128504da22677b53e695593de6efc89aae39, it's in the list.

mgascam reviewed (commented):
Thanks for working on this refactor @RadoslavGeorgiev. I appreciate it simplifies things. I don't have context on the reason for caching the payment method via `wp_options`, but the current approach using the user meta makes sense to me.

I saw you already solved comments from Leonardo so focused a bit more on manual testing. I followed the proposed instructions setting up the breakpoint and it worked as expected. I also tested:

- Adding a new payment method cleared the cache ✅ 
- Setting default payment method cleared the cache ✅
- Old cache entries are ignored and eventually cleaned ✅

Only left a couple of minor comments for confirmation.

mgascam reviewed (approved):
I intended to approve with my previous comment, but I accidentally selected Comment instead.

RadoslavGeorgiev committed:
Merge branch 'develop' into WOOPMNT-4879

RadoslavGeorgiev committed:
Add ADDRESS_AUTOCOMPLETE_JWT_KEY to the list of keys.
